src/Spice86.Core/Emulator/VM/CpuSpeedLimit/CpuCycleLimiter.cs

```diff
@@ -0,0 +1,87 @@
+﻿namespace Spice86.Core.Emulator.VM.CpuSpeedLimit;
+
+using Spice86.Core.Emulator.CPU;
+using static Spice86.Core.Emulator.Devices.Timer.Timer;
+
+using System.Diagnostics;
+
+/// <summary>
+/// Throttles CPU execution to maintain a target number of cycles per millisecond,
+/// using a budget-based approach similar to DOSBox.
+/// </summary>
+public class CpuCycleLimiter : CycleLimiterBase {
+    private readonly SpinWait _spinner = new();
+    private readonly Stopwatch _stopwatch = new();
+    private long _lastTicks;
+    private long _targetCyclesForPause;
+
+    private const int CyclesUp = 1000;
+    private const int CyclesDown = 1000;
+    private const int MaxCyclesPerMs = 60000;
+    private const int MinCyclesPerMs = 100;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CpuCycleLimiter"/> class.
+    /// </summary>
+    /// <param name="targetCpuCyclesPerMs">The target CPU cycles per millisecond. A value of 0 is corrected to 3000.</param>
+    public CpuCycleLimiter(int targetCpuCyclesPerMs) {
+        if (targetCpuCyclesPerMs > 0) {
+            TargetCpuCyclesPerMs = Math.Clamp(
+            targetCpuCyclesPerMs,
+            MinCyclesPerMs,
+            MaxCyclesPerMs);
+        } else {
+            TargetCpuCyclesPerMs = Math.Clamp(
+            ICyclesLimiter.RealModeCpuCyclesPerMs,
+            MinCyclesPerMs,
+            MaxCyclesPerMs);
+        }
+
+        _stopwatch.Start();
+    }
+
+    /// <inheritdoc/>
+    internal override void RegulateCycles(State cpuState) {
+        if (!cpuState.IsRunning) {
+            return;
+        }
+
+        // If current cycles haven't reached target yet, no need to regulate
+        if (cpuState.Cycles < _targetCyclesForPause) {
+            return;
+        }
+
+        // We've reached our target, time to regulate speed
+        long targetTicks = _lastTicks + StopwatchTicksPerMillisecond;
+
+        while (cpuState.IsRunning && _stopwatch.ElapsedTicks < targetTicks) {
+            _spinner.SpinOnce();
+        }
+
+        long wallClockTicks = _stopwatch.ElapsedTicks;
+
+        // (floating point for sub-millisecond precision)
+        double elapsedMs = (double)(wallClockTicks - _lastTicks) / Stopwatch.Frequency * 1000;
+
+        _lastTicks = wallClockTicks;
+
+        // Calculate how many cycles we should allow before the next pause
+        long cyclesToAdd = (long)(TargetCpuCyclesPerMs * elapsedMs);
```

Comment:
Not sure I understand the logic here.
If elapsedMs was actually 2ms (because PC was slower for example to execute 3000 instructions), it means we will allow 6000? Is my understanding correct? If so why?